src/sentry/relay/config/metric_extraction.py

```diff
@@ -1,26 +1,122 @@
 from dataclasses import dataclass
-from typing import Any, List, Literal, Sequence, Tuple, TypedDict, Union, cast
+from typing import Any, List, Literal, Optional, Sequence, Tuple, TypedDict, Union, cast
 
+from typing_extensions import NotRequired
+
+from sentry import features
 from sentry.api.endpoints.project_transaction_threshold import DEFAULT_THRESHOLD
+from sentry.constants import DataCategory
+from sentry.incidents.models import AlertRule, AlertRuleStatus
 from sentry.models import (
     Project,
     ProjectTransactionThreshold,
     ProjectTransactionThresholdOverride,
     TransactionMetric,
 )
 
+# GENERIC METRIC EXTRACTION
+
+
+# Name component of MRIs used for custom alert metrics.
+# TODO: Move to a common module.
+CUSTOM_ALERT_METRIC_NAME = "transactions/alert"
+
+# Version of the metric extraction config.
+_METRIC_EXTRACTION_VERSION = 1
+
 
-class RuleConditionInner(TypedDict):
-    op: Literal["eq", "gt", "gte"]
+RuleCondition = Union["LogicalRuleCondition", "ComparingRuleCondition", "NotRuleCondition"]
+
+
+class ComparingRuleCondition(TypedDict):
+    op: Literal["eq", "gt", "gte", "lt", "lte", "glob"]
     name: str
     value: Any
 
 
-# mypy does not support recursive types. type definition is a very small subset
-# of the values relay actually accepts
-class RuleCondition(TypedDict):
-    op: Literal["and"]
-    inner: Sequence[RuleConditionInner]
+class LogicalRuleCondition(TypedDict):
+    op: Literal["and", "or"]
+    inner: Sequence[RuleCondition]
+
+
+class NotRuleCondition(TypedDict):
+    op: Literal["not"]
+    inner: RuleCondition
+
+
+class TagSpec(TypedDict):
+    key: str
+    field: NotRequired[str]
+    value: NotRequired[str]
+    condition: NotRequired[RuleCondition]
+
+
+class MetricSpec(TypedDict):
+    category: Literal["transaction"]
+    mri: str
+    field: NotRequired[Optional[str]]
+    condition: NotRequired[RuleCondition]
+    tags: NotRequired[Sequence[TagSpec]]
+
+
+class MetricExtractionConfig(TypedDict):
+    version: int
+    metrics: List[MetricSpec]
+
+
+def get_metric_extraction_config(project: Project) -> Optional[MetricExtractionConfig]:
+    if not features.has("organizations:metrics-extraction", project.organization):
+        return None
+
+    alerts = AlertRule.objects.filter(
+        project=project, status=AlertRuleStatus.PENDING
+    ).select_related("snuba_query")
+
+    metrics: List[MetricSpec] = []
+    for alert in alerts:
```

Comment:
This will create the same metric for each pending alert (hardcoded in `_convert_alert_to_metric`). Considering it's the same one and that we don't know the number of pending metrics, what do you think about limiting this number? I suggest to a) not add duplicates, and b) set an arbitrary maximum to 20(?) to not grow the size of a project config too much.

Reply:
These are all valid concerns but we are not working for a GA implementation, we are still in exploratory phase.

Limiting and de-duplication of metrics is something that we will for sure take a look at, since there is a high chance (especially for widgets) that there will be identical metrics.

Reply:
This is a concern that affects the reliability of the infrastructure. We discussed this and follow-ups on Slack on the project channel.

Reply:
I'll add a first conservative limit, but be advised that this state is supposed to be inactive. There's no point in enabling this until a more complete implementation is added.